bee-network/src/service/service.rs

```diff
@@ -0,0 +1,486 @@
+                peerlist,
+            } = config;
+
+            let Receivers {
+                commands,
+                internal_events,
+            } = receivers;
+
+            let (shutdown_tx1, shutdown_rx1) = oneshot::channel::<()>();
+            let (shutdown_tx2, shutdown_rx2) = oneshot::channel::<()>();
+            let (shutdown_tx3, shutdown_rx3) = oneshot::channel::<()>();
+
+            tokio::spawn(async move {
+                shutdown.await.expect("receiving shutdown signal");
+
+                shutdown_tx1.send(()).expect("receiving shutdown signal");
+                shutdown_tx2.send(()).expect("receiving shutdown signal");
+                shutdown_tx3.send(()).expect("receiving shutdown signal");
+            });
+            tokio::spawn(command_processor(
+                shutdown_rx1,
+                commands,
+                senders.clone(),
+                peerlist.clone(),
+            ));
+            tokio::spawn(event_processor(
+                shutdown_rx2,
+                internal_events,
+                senders.clone(),
+                peerlist.clone(),
+            ));
+            tokio::spawn(peer_checker(shutdown_rx3, senders, peerlist));
+
+            info!("Network service started.");
+        }
+    }
+}
+
+async fn command_processor(shutdown: Shutdown, commands: CommandReceiver, senders: Senders, peerlist: PeerList) {
+    debug!("Command handler running.");
+
+    let mut commands = ShutdownStream::new(shutdown, UnboundedReceiverStream::new(commands));
+
+    while let Some(command) = commands.next().await {
+        if let Err(e) = process_command(command, &senders, &peerlist).await {
+            error!("Error processing command. Cause: {}", e);
+            continue;
+        }
+    }
+
+    debug!("Command processor stopped.");
+}
+
+async fn event_processor(shutdown: Shutdown, events: InternalEventReceiver, senders: Senders, peerlist: PeerList) {
+    debug!("Event handler running.");
+
+    let mut int_events = ShutdownStream::new(shutdown, UnboundedReceiverStream::new(events));
+
+    while let Some(int_event) = int_events.next().await {
+        if let Err(e) = process_internal_event(int_event, &senders, &peerlist).await {
+            error!("Error processing internal event. Cause: {}", e);
+            continue;
+        }
+    }
+
+    debug!("Event processor stopped.");
+}
+
+// TODO: implement exponential back-off to not spam the peer with reconnect attempts.
+async fn peer_checker(shutdown: Shutdown, senders: Senders, peerlist: PeerList) {
+    debug!("Peer checker running.");
+
+    let Senders { internal_commands, .. } = senders;
+
+    // NOTE:
+    // We add a random amount of milliseconds to when the reconnector starts, so that even if 2 nodes
+    // go online at the same time the probablilty of them simultaneously dialing each other is reduced
+    // significantly.
+
+    let delay = Duration::from_millis(rand::thread_rng().gen_range(0u64..1000));
```

Comment:
Might be worth documenting the tradeoff that's going on here. At the upper bound of the range increases, the average delay increases linearly, but the chance of a mutual dial reduces quadratically (I think?).

Reply:
So the way that this works is as such:
Let's say two nodes `A` and `B` start at the same time `t0`. Both individually pick a random number within the range `[0...1000)` (representing milliseconds). Then `A` will dial `B` at `t0 + delay(A)` and `B` at `t0 + delay(B)`. This is IMHO sufficient to prevent mutual dialing. There is in any case only a very small probability anyway that this becomes an issue. The `reconnect_interval_secs` remains the same for all nodes. So let's say 2 nodes lose their connection, then they don't immediatedly try to reconnect to eachother, but instead wait for the PeerState checker to initiate reconnection.

Reply:
I think that sounds good, would be nice to see that in the code.

Reply:
Done